url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/Eval.java

```
@@ -0,0 +1,762 @@
+ *    <li>Only strings are provided, there is no 'symbol' concept. If an application wants e.g. numbers
+ *        for calculation, the commands should parse the strings.</li>
+ *    <li>Binding context lookup order. Scoping is lexical. The global context is always available for everybody,
+ *        there is no need to introduce dynamic scoping.
+ *        <ol>
+ *           <li>Call context, arguments are bound to parameters. This context is especially created for this call. It contains all local bindings.</li>
+ *           <li>Lexical (static) context, where the function or lambda was defined. It is the closure of the lambda.</li>
+ *        </ol></li>
+ *     <li>Interrupting the eval thread will lead to a CommandException. This is the preferred way to stop a runaway script.</li>
+ * </ul>
+ *
+ */
+public class Eval
+extends AbstractEval
+{
+    private CommandRepository commands;
+    private CommandRepository macros;
+
+    public Eval()
+    {
+        this(new BasicContext());
+    }
+
+    public Eval(IContext aContext)
+    {
+        super(aContext);
+        commands = new CommandRepository();
+        macros = new CommandRepository();
+    }
+
+    public void setCommandRepo(CommandRepository aRepo)
+    {
+        commands = aRepo;
+    }
+
+    public CommandRepository getCommandRepo()
+    {
+        return commands;
+    }
+
+    public void setMacroRepo(CommandRepository aRepo)
+    {
+       macros = aRepo;
+    }
+
+    public CommandRepository getMacroRepo()
+    {
+        return macros;
+    }
+
+    /**
+     * Evaluate an expression in the root REPL context.
+     *
+     * @param aExpr An expression.
+     * @return The result of evaluation.
+     * @throws CommandException
+     */
+    public Object eval(Object aExpr)
+    throws CommandException
+    {
+        try
+        {
+            return eval(aExpr, getContext());
+        }
+        catch (StackOverflowError e)
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)